src/code42cli/logger/handlers.py

```diff
@@ -0,0 +1,145 @@
+    beginning of the message. Most CEF consumers seem to not expect the `<PRI>` to be
+    present in CEF messages. Attach to a logger via `.addHandler` to use.
+
+    `self.socket` is lazily loaded for testing purposes, so the connection does not get
+    made for TCP/TLS until the first log record is about to be transmitted.
+
+    Args:
+        hostname: The hostname of the syslog server to send log messages to.
+        port: The port of the syslog server to send log messages to.
+        protocol: The protocol over which to submit syslog messages. Accepts TCP, UDP, or TLS.
+        certs: Certs to specify when using TLS-TCP for the `protocol` argument. Use "ignore" for
+            ssl.CERT_NONE (ignoring certificate validation).
+    """
+
+    def __init__(self, hostname, port, protocol, certs):
+        self._hostname = hostname
+        self._port = port
+        self._protocol = protocol
+        self._certs = certs
+        self.address = (hostname, port)
+        logging.Handler.__init__(self)
+        self.socktype = _try_get_socket_type_from_protocol(protocol)
+        self.socket = None
+
+    @property
+    def _wrap_socket(self):
+        return self._protocol == ServerProtocol.TLS_TCP
+
+    def connect_socket(self):
+        """Call to initialize the socket. If using TCP/TLS, it will also establish the connection.
+        """
+        if not self.socket:
+            self.socket = self._create_socket(self._hostname, self._port, self._certs)
+
+    def _create_socket(self, hostname, port, certs):
+        socket_info = self._get_socket_address_info(hostname, port)
+        address_family, sock_type, proto, _, sa = socket_info
+        sock = None
+        try:
+            sock = socket.socket(address_family, sock_type, proto)
+            if self._wrap_socket:
+                sock = _wrap_socket_for_ssl(sock, certs, hostname)
+            if sock_type == socket.SOCK_STREAM:
+                sock = _connect_socket(sock, sa)
+            return sock
+        except Exception as exc:
+            if sock is not None:
+                sock.close()
+            raise exc
+
+    def _get_socket_address_info(self, hostname, port):
+        info = socket.getaddrinfo(hostname, port, 0, self.socktype)
+        if not info:
+            raise OSError("getaddrinfo() returns an empty list")
+        return info[0]
+
+    def emit(self, record):
+        try:
+            self._send_record(record)
+        except Exception:
+            self.handleError(record)
+
+    def handleError(self, record):
+        """Override logger's `handleError` method to exit if an exception is raised while trying to
+        log, otherwise it would continue to gather and process events if the connection breaks but send
+        them nowhere.
+        """
+        t, _, _ = sys.exc_info()
+        if t == BrokenPipeError:
+            raise SyslogServerNetworkConnectionError()
```

Comment:
To get `Code42CLIError` to work, we raise a custom exception here, and that gets caught in `click_ext.groups.py` line 61, which then gets converted into a `Code42CLIError`. The reason for this is because we are unable to import `Code42CLIError` here because of a circular dependency.  Before, it was raising a `ClickException`, but that text is not red.